### PR TITLE
file renames in post action arguments PoC

### PR DESCRIPTION
**Pull request by vlada-shubina**
### Problem
https://github.com/dotnet/templating/issues/1600
unblocking https://github.com/dotnet/sdk/pull/29880

### Solution
Added ability to apply file renames to post action arguments

TODOs:
- [x] validation
- [x] tests
- [x] documentation update
- [x] nth: `FileRenameGenerator` refactoring (processor can be setup once for all primary outputs and post actions replacements)

### Checks:
- [x] Added unit tests
- [x] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

### CODE REVIEW

**Review by YuliiaKovalova — APPROVED** [review]

File: test/Microsoft.TemplateEngine.TestTemplates/test_templates/PostActions/WithFileRename/.template.config/template.json [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    {
+      "description": "Modify JSON file",
+      "manualInstructions": [ { "text": "Add MyTestProject property to testfile.json manually." } ],
+      "args": {
+        "jsonFileName": "testfile.json",
+        "parentPropertyPath": "moduleConfiguration:edgeAgent:properties.desired:modules",
+        "newJsonPropertyName": "MyTestProject",
+        "newJsonPropertyValue": "${MODULEDIR<../MyTestProject>}"
+      },
+      "applyFileRenamesToArgs": [ "newJsonPropertyName", "newJsonPropertyValue" ],
```
Comment (fgheysels) [inline test/Microsoft.TemplateEngine.TestTemplates/test_templates/PostActions/WithFileRename/.template.config/template.json:28]: I wonder if `applyFileRenamesToArgs` is not a bit confusing.  You're actually not renaming any files ?
Maybe `applySourceRenamesToArgs` is better ? (Or even simply plain `applyRenamesToArgs`? )

File: test/Microsoft.TemplateEngine.IDE.IntegrationTests/End2EndTests.cs [hunk 2]
```diff
@@ -423,6 +424,33 @@ internal async Task Test_CreateAsync_PreferDefaultNameInvalidParameters()
...
+            ITemplateCreationResult result = await bootstrapper
+                .CreateAsync(foundTemplates[0].Info, "CompanyProject", output, parameters)
+                .ConfigureAwait(false);
+
+            IPostAction postAction = Assert.Single(result.CreationResult!.PostActions);
+            Assert.Equal("testfile.json", postAction.Args["jsonFileName"]);
+            Assert.Equal("moduleConfiguration:edgeAgent:properties.desired:modules", postAction.Args["parentPropertyPath"]);
+            Assert.Equal("CompanyProject", postAction.Args["newJsonPropertyName"]);
+            Assert.Equal("${MODULEDIR<../CompanyProject>}", postAction.Args["newJsonPropertyValue"]);
+            Assert.Equal("Add CompanyProject property to testfile.json manually.", postAction.ManualInstructions);
```
Comment (fgheysels) [inline test/Microsoft.TemplateEngine.IDE.IntegrationTests/End2EndTests.cs:451]: When you use t he modified args in the PostAction (in this case the `newJsonPropertyName` for instance, will this change be available in the custom action then ? ).

I mean, I suppose that the renames to those args occur before the code of the custom PostAction is executed ?

File: test/Microsoft.TemplateEngine.TestTemplates/test_templates/PostActions/WithFileRename/.template.config/template.json (hunk 1 above)
Comment (vlada-shubina) [reply test/Microsoft.TemplateEngine.TestTemplates/test_templates/PostActions/WithFileRename/.template.config/template.json:28]: i used `fileRename`, as the configuration for symbol that is used is called `fileRename`, please check another example here: https://github.com/vlada-shubina/templating/blob/4c2bab980aa1f93ed00186efa6376780b08fb6b8/test/Microsoft.TemplateEngine.Orchestrator.RunnableProjects.UnitTests/TemplateConfigTests/PostActionTests.cs#L272-L307. 

`sourceName` is a special symbol, for which both `replaces` and `fileRename` value is same and corresponds to value of `sourceName`. 

We can definitely shorten it to `applyRenamesToArgs`. I would not recommend using `applySourceRenamesToArg` which may imply using [source file renames](https://github.com/dotnet/templating/wiki/Reference-for-template.json#source-definition), which are not applied in this scenario. Only symbol based renames are applied.

Comment (fgheysels) [reply test/Microsoft.TemplateEngine.TestTemplates/test_templates/PostActions/WithFileRename/.template.config/template.json:28]: Well, it was just a suggestion as the first time I saw the name, I thought that filenames were going to be renamed. :)

**Review by fgheysels — APPROVED** [review]

**Pull request merged**